examples/autoqasm/3_2_magic_state_distillation.ipynb

```diff
@@ -0,0 +1,447 @@
+   "id": "73dd3d78-1444-4520-98cc-dad077a97d36",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@aq.main(num_qubits=5)\n",
+    "def distillation():\n",
+    "    qubits = range(5)\n",
+    "\n",
+    "    # state preparation\n",
+    "    for q in aq.ArrayVar(qubits, dimensions=[5]):\n",
+    "        physical_magic_state(q)\n",
+    "\n",
+    "    # decoding\n",
+    "    decoder(*qubits)\n",
+    "\n",
+    "    # final rotation\n",
+    "    ins.h(qubits[0])\n",
+    "    ins.y(qubits[0])\n",
+    "\n",
+    "    # measure ancilla\n",
+    "    c = ins.measure(qubits[1:5])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f3f2b76e-5996-410b-a53f-4196ccc9df6f",
+   "metadata": {},
+   "source": [
+    "We run the distillation program on a local simulator."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "43f4c1b1-dbcc-4ee0-8f99-db087bb0c8d8",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "measurement counts:  Counter({'0000': 155, '1100': 69, '1001': 60, '0010': 60, '0110': 60, '1011': 59, '0011': 58, '0001': 58, '1000': 58, '1111': 56, '0101': 55, '0100': 55, '1110': 53, '0111': 50, '1101': 48, '1010': 46})\n"
+     ]
+    }
+   ],
+   "source": [
+    "n_shots = 1000\n",
+    "result = LocalSimulator().run(distillation(), shots=n_shots).result()\n",
+    "counts = Counter(result.measurements[\"c\"])\n",
+    "print(\"measurement counts: \", counts)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8126350f-c7bd-4ec9-bbdb-74a9300fc3d9",
+   "metadata": {},
+   "source": [
+    "About 1/6 of the total shots correspond to successful preparations of magic state on qubit 0 (i.e., measuring \"0000\" in ancilla), agreeing with the result in Ref[1]."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "c0e78ced-1137-4d6e-be88-2cc90b40105d",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "success ratio:  0.155\n"
+     ]
+    }
+   ],
+   "source": [
+    "success_count = len([x for x in result.measurements[\"c\"] if x==\"0000\"])\n",
+    "print(\"success ratio: \", success_count/n_shots)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "74c6e629-48fc-4310-bbac-7194f9c6806c",
+   "metadata": {},
+   "source": [
+    "## Repeat until success\n",
+    "To guarantee successful preparation of magic state in every shot, we introduce a repeat-until-success (RUS) protocol [3]. The state preparation and the decoding are repeated until all ancilla is measured in \"0\". In this way, non-successful outcomes are discarded before starting over. It guarantees that every shot ends up with successful magic state preparation, distilling the magic state. This protocol requires the expressibility of while-loop in the quantum program because the looping condition is based on the result of each iteration."
```

Comment:
```suggestion
    "To guarantee a successful preparation of the magic state in every shot, we introduce a repeat-until-success (RUS) protocol [3]. The state preparation and the decoding are repeated until all ancilla measure \"0\". This protocol requires `while`-loops in the quantum program because we continue to retry until the success condition is met."
```
Not sure the middle two sentences add much info, I think it's already pretty clear. But it's not a strong suggestion